apache2/msc_multipart.c

```diff
@@ -198,23 +198,44 @@ static int multipart_parse_content_disposition(modsec_rec *msr, char *c_d_value)
                     log_escape_nq(msr->mp, value));
             }
         }
-        else
-        if (strcmp(name, "filename") == 0) {
+        else if ((strcmp(name, "filename") == 0) || (strcmp(name, "filename*") == 0))
+        {
 
-            validate_quotes(msr, value);
+            char *decoded_filename = NULL;
 
-            msr->multipart_filename = apr_pstrdup(msr->mp, value);
+            if (strcmp(name, "filename*") == 0)
+            {
+                decoded_filename = rfc5987_decode(msr->mp, value);
+                if (!decoded_filename)
+                {
+                    msr_log(msr, 4, "Multipart: Could not decode extended filename parameter in RFC 5987 format: %s",
+                            log_escape_nq(msr->mp, value));
+                    return -16;
+                }
+                msr->multipart_filename = decoded_filename;
+
+                // Make sure to turn of INVALID quoting since RFC 5987 expects quotes in the filename format.
+                msr->mpd->flag_invalid_quoting = 0;
```

Comment:
would it be smarter to fix the code that decided that the quoting was invalid? Or maybe we are not even running validate_quotes now? Shouldn't we still be running validate_quotes?

Reply:
the place where the quotes are checked are when parsing the content disposition header, the `name` field and the `filename` field (there seems to be some code duplication here as well). So what I could gather was that Modsecurity doesn't expect quotes in `name` or the `filename` fields which makes sense. So I ended making this change to just make sure that there is an exception to this rule when we are dealing with RFC 5987 encoded string.